Realm/RLMLogger.h

```diff
@@ -0,0 +1,129 @@
+////////////////////////////////////////////////////////////////////////////
+//
+// Copyright 2023 Realm Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+////////////////////////////////////////////////////////////////////////////
+
+#import <Realm/RLMConstants.h>
+
+RLM_HEADER_AUDIT_BEGIN(nullability)
+
+/// An enum representing different levels of sync-related logging that can be configured.
+typedef RLM_CLOSED_ENUM(NSUInteger, RLMLogLevel) {
+    /// Nothing will ever be logged.
+    RLMLogLevelOff,
+    /// Only fatal errors will be logged.
+    RLMLogLevelFatal,
+    /// Only errors will be logged.
+    RLMLogLevelError,
+    /// Warnings and errors will be logged.
+    RLMLogLevelWarn,
+    /// Information about sync events will be logged. Fewer events will be logged in order to avoid overhead.
+    RLMLogLevelInfo,
+    /// Information about sync events will be logged. More events will be logged than with `RLMSyncLogLevelInfo`.
+    RLMLogLevelDetail,
+    /// Log information that can aid in debugging.
+    ///
+    /// - warning: Will incur a measurable performance impact.
+    RLMLogLevelDebug,
+    /// Log information that can aid in debugging. More events will be logged than with `RLMSyncLogLevelDebug`.
+    ///
+    /// - warning: Will incur a measurable performance impact.
+    RLMLogLevelTrace,
+    /// Log information that can aid in debugging. More events will be logged than with `RLMSyncLogLevelTrace`.
+    ///
+    /// - warning: Will incur a measurable performance impact.
+    RLMLogLevelAll
+} NS_SWIFT_NAME(LogLevel);
+
+/**
+ `RLMLogger` is a base class for creating your own custom logging logic.
```

Comment:
Why does this API involve defining a subclass rather than just a logging callback like the old API? I don't really see why we'd want to expose an interface for users to do their own logging via the logger they passed to us, and the base class doesn't add any flexibility.

Reply:
I think the minimal reasonable API we could do which doesn't lock us into just a single global default logger would be something like:

```swift
class Logger {
    public class var shared: Logger
    public var level: LogLevel
    public init(level: LogLevel, logFunction: (LogLevel, String) -> Void)
}

// Used as
Logger.shared = Logger(level: .info) { level, msg in
  print("\(level): \(msg)")
}
Logger.shared.level = .warn